pkg/cluster/spec/tiflash.go

```diff
@@ -640,25 +641,36 @@ type replicateConfig struct {
 	EnablePlacementRules string `json:"enable-placement-rules"`
 }
 
-func (i *TiFlashInstance) getEndpoints() []string {
+func (i *TiFlashInstance) getEndpoints(topo Topology) []string {
 	var endpoints []string
-	for _, pd := range i.topo.(*Specification).PDServers {
+	for _, pd := range topo.(*Specification).PDServers {
 		endpoints = append(endpoints, fmt.Sprintf("%s:%d", pd.Host, uint64(pd.ClientPort)))
 	}
 	return endpoints
 }
 
 // PrepareStart checks TiFlash requirements before starting
-func (i *TiFlashInstance) PrepareStart(tlsCfg *tls.Config) error {
-	endPoints := i.getEndpoints()
+func (i *TiFlashInstance) PrepareStart(ctx context.Context, tlsCfg *tls.Config) error {
 	// set enable-placement-rules to true via PDClient
-	pdClient := api.NewPDClient(endPoints, 10*time.Second, tlsCfg)
 	enablePlacementRules, err := json.Marshal(replicateConfig{
 		EnablePlacementRules: "true",
 	})
+	// this should not failed, else exit
 	if err != nil {
-		return nil
+		return err
+	}
+
+	topo := ctx.Value(ctxt.CtxBaseTopo)
+	if topo == nil {
+		topo = i.topo
+	}
+
+	topo, ok := topo.(Topology)
+	if !ok {
+		return perrs.New("base topology in context is invalid")
 	}
 
+	endpoints := i.getEndpoints(topo.(Topology))
```

Comment:
```suggestion
	endpoints := i.getEndpoints(topo)
```

Reply:
Sorry, the variable `topo` was firstly introduced in line 663, that is the value of context, its type is `interface{}`,  the line 668 short variable declaration doesn't change its type.

I'll change the code style above to make it more maintainable.